keyboards/ploopyco/madromys/madromys.c

```diff
@@ -0,0 +1,190 @@
+/* Copyright 2023 Colin Lam (Ploopy Corporation)
+ * Copyright 2020 Christopher Courtney, aka Drashna Jael're  (@drashna) <drashna@live.com>
+ * Copyright 2019 Sunjun Kim
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "madromys.h"
+
+#ifndef PLOOPY_DPI_OPTIONS
+#    define PLOOPY_DPI_OPTIONS \
+        { 600, 900, 1200, 1600 }
+#    ifndef PLOOPY_DPI_DEFAULT
+#        define PLOOPY_DPI_DEFAULT 1
+#    endif
+#endif
+#ifndef PLOOPY_DPI_DEFAULT
+#    define PLOOPY_DPI_DEFAULT 0
+#endif
+#ifndef PLOOPY_DRAGSCROLL_DPI
+#    define PLOOPY_DRAGSCROLL_DPI 100  // Fixed-DPI Drag Scroll
+#endif
+#ifndef PLOOPY_DRAGSCROLL_FIXED
+#    define PLOOPY_DRAGSCROLL_FIXED 1
+#endif
+#ifndef PLOOPY_DRAGSCROLL_MULTIPLIER
+#    define PLOOPY_DRAGSCROLL_MULTIPLIER 0.75  // Variable-DPI Drag Scroll
+#endif
+
+keyboard_config_t keyboard_config;
+uint16_t          dpi_array[] = PLOOPY_DPI_OPTIONS;
+#define DPI_OPTION_SIZE (sizeof(dpi_array) / sizeof(uint16_t))
+
+// TODO: Implement libinput profiles
+// https://wayland.freedesktop.org/libinput/doc/latest/pointer-acceleration.html
+// Compile time accel selection
+// Valid options are ACC_NONE, ACC_LINEAR, ACC_CUSTOM, ACC_QUADRATIC
+
+// Trackball State
+bool     is_scroll_clicked = false;
+uint16_t last_scroll       = 0;  // Previous confirmed wheel event
+uint16_t last_mid_click    = 0;  // Stops scrollwheel from being read if it was pressed;
+bool     debug_encoder     = false;
+bool     is_drag_scroll    = false;
+
+// require, since core encoder.c (where is is normally defined isn't present
+__attribute__((weak)) bool encoder_update_user(uint8_t index, bool clockwise) { return true; }
+
+bool encoder_update_kb(uint8_t index, bool clockwise) {
+    if (!encoder_update_user(index, clockwise)) {
+        return false;
+    }
+#ifdef MOUSEKEY_ENABLE
+    tap_code(clockwise ? KC_WH_U : KC_WH_D);
+#else
+    report_mouse_t mouse_report = pointing_device_get_report();
+    mouse_report.v              = clockwise ? 1 : -1;
+    pointing_device_set_report(mouse_report);
+    pointing_device_send();
+#endif
+    return true;
+}
+
+void encoder_init(void) {
+    return;
+}
+
+bool encoder_read(void) {
+    return false;
+}
```

Comment:
no encoder?
```suggestion
// Trackball State
bool     debug_encoder     = false;
bool     is_drag_scroll    = false;

```

Reply:
Committed partially in [41a834f](https://github.com/qmk/qmk_firmware/pull/21989/commits/41a834f9b1d993a40a52e9ae7e616de1dc879e84). `encoder_init` and `encoder_read` were left in because compilation failed with them removed:

    Linking: .build/ploopyco_madromys_rev1_001_default.elf                                              [ERRORS]
     |
     | c:/qmk_msys/mingw64/bin/../lib/gcc/arm-none-eabi/10.1.0/../../../../arm-none-eabi/bin/ld.exe: .build/obj_ploopyco_madromys_rev1_001_default/quantum/keyboard.o: in function `keyboard_init':
     | C:\Users\Colin\Ploopy\github\qmk-firmware-ploopyco-dev/quantum/keyboard.c:417: undefined reference to `encoder_init'
     | c:/qmk_msys/mingw64/bin/../lib/gcc/arm-none-eabi/10.1.0/../../../../arm-none-eabi/bin/ld.exe: .build/obj_ploopyco_madromys_rev1_001_default/quantum/keyboard.o: in function `keyboard_task':
     | C:\Users\Colin\Ploopy\github\qmk-firmware-ploopyco-dev/quantum/keyboard.c:673: undefined reference to `encoder_read'
     | collect2.exe: error: ld returned 1 exit status
     |

Putting them back fixed that.

`is_drag_scroll` was left in because drag scroll breaks without it.